### PR TITLE
Folder metadata in stat cache

### PR DESCRIPTION
**Pull request by ankitaluthra1**
### Description
Adds insertFolder, AddNegativeEntryForFolder and lookup folder in stat cache for HNS buckets.
This PR only adds new methods with corresponding implementations, currently these are not called from any bucket type.
### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Done
3. Integration tests - NA

### CODE REVIEW

File: internal/cache/metadata/stat_cache.go [hunk 1]
```diff
@@ -203,3 +219,46 @@ func (sc *statCacheBucketView) LookUp(
 
 	return
 }
+
+func (sc *statCacheBucketView) LookUpFolder(
+	folderName string,
+	now time.Time) (bool, *controlpb.Folder) {
+	// Look up in the LRU cache.
+	value := sc.sharedCache.LookUp(sc.key(folderName))
```
Comment (Tulsishah) [inline internal/cache/metadata/stat_cache.go:227]: Should we create common lookUp function which will return entry and then on respective functions we will assign folder or minObject?

Same for insert.

File: internal/cache/metadata/stat_cache.go [hunk 2]
```diff
@@ -203,3 +219,46 @@ func (sc *statCacheBucketView) LookUp(
 
 	return
 }
+
```
Comment (Tulsishah) [inline internal/cache/metadata/stat_cache.go:222]: AddNegativeEntry function will require to set f to be nil

Comment (Tulsishah) [inline internal/cache/metadata/stat_cache.go:222]: Size()  function will require to add size of e.f.

Not sure if you want to make changes in next PR for these

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go:222]: negative entry and size are part of next PR, trying to make smallest PR for quicker reviews

Comment (Tulsishah) [reply internal/cache/metadata/stat_cache.go:222]: 2 line changes I guess required for rest of the two functions, I guess we can add that in this PR so all the stat cache related changes got merged in single PR.

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go:222]: it wont be just 2 lines, new unit tests are also needed to test negative case

Comment (Tulsishah) [reply internal/cache/metadata/stat_cache.go:222]: Oh okay sorry forgot about unit tests.

File: internal/cache/metadata/stat_cache_test.go [hunk 3]
```diff
@@ -408,3 +411,72 @@ func (t *MultiBucketStatCacheTest) Test_ExpiresLeastRecentlyUsed() {
...
+
+	t.statCache.InsertFolder(newEntry, expiration)
+
+	hit, entry := t.statCache.LookUpFolder(name, someTime)
+	assert.True(t.T(), hit)
+	assert.Equal(t.T(), "key1", entry.Name)
+	assert.Equal(t.T(), int64(2), entry.Metageneration)
+}
+
+func (t *StatCacheTest) Test_Lookup_Return_False_If_Expiration_Is_Passed() {
```
Comment (Tulsishah) [inline internal/cache/metadata/stat_cache_test.go:450]: Should we write unit tests for entry equals to nil?

File: internal/cache/metadata/stat_cache.go (hunk 1 above)
Comment (Tulsishah) [reply internal/cache/metadata/stat_cache.go:227]: e.g.
```
lookUp(return entry)
lookUpObject(set object from entry)
lookUpFolder(set folder from entry)
```
If it takes more effort then we can skip may be.

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: Done for lookup, Insert is not having much repetitive code for now and so not refactoring insert for now.

File: internal/cache/metadata/stat_cache_test.go (hunk 3 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:450]: added

File: internal/cache/metadata/stat_cache_test.go [hunk 4]
```diff
@@ -111,6 +113,7 @@ type MultiBucketStatCacheTest struct {
 func (t *StatCacheTest) SetupTest() {
 	cache := lru.NewCache(uint64((mount.AverageSizeOfPositiveStatCacheEntry + mount.AverageSizeOfNegativeStatCacheEntry) * capacity))
 	t.cache.wrapped = metadata.NewStatCacheBucketView(cache, "") // this demonstrates
+	t.statCache = metadata.NewStatCacheBucketView(cache, "")     // this demonstrates
```
Comment (Tulsishah) [inline internal/cache/metadata/stat_cache_test.go:116]: can't we use `t.cache` only? why did we need this separate variable?

**Review by Tulsishah — COMMENTED** [review]
LGTM

File: internal/cache/metadata/stat_cache_test.go (hunk 4 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:116]: t.cache is the old wrapper class in tests that calls metadata.StatCache. This approach tests wrapper methods instead of actual functionality, compromising the safety net. If the helper class changes internally, tests won't fail. I've refactored this in my PR to test metadata.StatCache directly, removing unnecessary wrappers for accurate unit testing. Changing every test will increase the scope and actual hns work will be affected, so taking cautious call to just add new test in refactored way in first go, we can update rest of tests slowly later.

Comment (Tulsishah) [reply internal/cache/metadata/stat_cache_test.go:116]: Okay just write comment above it.

**Review by Tulsishah — APPROVED** [review]

File: internal/cache/metadata/stat_cache_test.go [hunk 5]
```diff
@@ -408,3 +411,128 @@ func (t *MultiBucketStatCacheTest) Test_ExpiresLeastRecentlyUsed() {
...
+
+	assert.True(t.T(), hit1)
+	assert.Equal(t.T(), "1", entry1.Name)
+	assert.True(t.T(), hit2)
+	assert.Equal(t.T(), "2", entry2.Name)
+
+	t.statCache.InsertFolder(folderEntry, expiration) //adds size of 220 and exceeds capacity
+
+	hit1, entry1 = t.statCache.LookUp("1", someTime)
+	hit3, entry3 := t.statCache.LookUpFolder("3", someTime)
```
Comment (Tulsishah) [inline internal/cache/metadata/stat_cache_test.go:531]: If I am correct It will cache now objectEntry1  and folderEntry and evict objectEntry2, so size function should return `1428 + 220 = 1648` if so Let's make assert for this.  

Also should we make block of setup, then calls size function(block2) and the assert of size(block3)?

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:531]: > If I am correct It will cache now objectEntry1 and folderEntry and evict objectEntry2, so size function should return 1428 + 220 = 1648 if so Let's make assert for this.
assertions for entry 1 and 2 are already in test, entry 1 is evicted, assertions are added for the same and not 2, least recently used is entry 1

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:531]: I have added blocks based on sequence of testing, since this test is testing two scenarios in one test, AAA for first scenario - where two object entries are added and then AAA for second scenario - in which folder entry is added

Comment (Tulsishah) [reply internal/cache/metadata/stat_cache_test.go:531]: Got it thanks

**Review by Tulsishah — DISMISSED** [review]

**Review by Tulsishah — DISMISSED** [review]

File: internal/cache/metadata/stat_cache.go [hunk 6]
```diff
@@ -176,30 +201,86 @@ func (sc *statCacheBucketView) AddNegativeEntry(objectName string, expiration ti
...
+	// Look up in the LRU cache.
+	entry := sc.lruCacheLookup(folderName, now)
+	if entry == nil {
+		return false, nil
+	}
+
+	return true, entry.f
+}
+
+func (sc *statCacheBucketView) lruCacheLookup(key string, now time.Time) *entry {
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:248]: can you rename this to cachelookup or sharedcachelookup.
too many names will be confusing - method is called lrucache , inside we are using sharedcache.

Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:248]: how about returning a bool here which says entry is found or not.

File: internal/cache/metadata/stat_cache_test.go [hunk 7]
```diff
@@ -408,3 +418,128 @@ func (t *MultiBucketStatCacheTest) Test_ExpiresLeastRecentlyUsed() {
 	assert.Equal(t.T(), cardamom, spices.LookUpOrNil("cardamom", someTime))
 	assert.Equal(t.T(), saffron, spices.LookUpOrNil("saffron", someTime))
 }
+
+func (t *StatCacheTest) Test_Create_Entry_When_No_Entry_Is_Present() {
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache_test.go:422]: naming of test methods

Test_methodNameUnderTest_conditionAndExpectedResult or
methodNameUnderTest_condition_expectedResult

See what is followed in gcsfuse and name appropriately.

File: internal/cache/metadata/stat_cache_test.go [hunk 8]
```diff
@@ -408,3 +418,128 @@ func (t *MultiBucketStatCacheTest) Test_ExpiresLeastRecentlyUsed() {
...
+	assert.Equal(t.T(), int64(1), entry.Metageneration)
+}
+
+func (t *StatCacheTest) Test_Override_Entry_Old_Entry_Is_Already_Present() {
+	const name = "key1"
+	existingEntry := &controlpb.Folder{
+		Name:           name,
+		Metageneration: 1,
+	}
+	t.statCache.InsertFolder(existingEntry, expiration)
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache_test.go:443]: check the metageneration of existing entry.

File: internal/cache/metadata/stat_cache_test.go [hunk 9]
```diff
@@ -408,3 +418,128 @@ func (t *MultiBucketStatCacheTest) Test_ExpiresLeastRecentlyUsed() {
...
+
+	assert.True(t.T(), hit1)
+	assert.Equal(t.T(), "1", entry1.Name)
+	assert.True(t.T(), hit2)
+	assert.Equal(t.T(), "2", entry2.Name)
+
+	t.statCache.InsertFolder(folderEntry, expiration) //adds size of 220 and exceeds capacity
+
+	hit1, entry1 = t.statCache.LookUp("1", someTime)
+	hit3, entry3 := t.statCache.LookUpFolder("3", someTime)
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache_test.go:538]: can we also check if entry2 is still present.

File: internal/cache/metadata/stat_cache.go [hunk 10]
```diff
@@ -92,14 +113,18 @@ type entry struct {
...
+	// First, calculate size on heap (including folder size also in case of hns buckets, in case of non-hns buckets 0 will be added as e.f will be Nil ).
 	// Additional 2*util.UnsafeSizeOf(&e.key) is to account for the copies of string
 	// struct stored in the cache map and in the cache linked-list.
 	size = uint64(util.UnsafeSizeOf(&e) + len(e.key) + 2*util.UnsafeSizeOf(&e.key) + util.NestedSizeOfGcsMinObject(e.m))
 	if e.m != nil {
 		size += 515
 	}
 
+	if e.f != nil {
+		size = size + uint64(util.UnsafeSizeOf(&e.f))
```
[the remark this answers is not in the document]
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: done

File: internal/cache/metadata/stat_cache.go (hunk 6 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: I also had same dilemma to return additional bool or not then I thought its redundant information based entry returned caller can find out if cache is a hit or miss.

File: internal/cache/metadata/stat_cache_test.go (hunk 7 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go]: added Test_methodNameUnderTest_ as per convention

File: internal/cache/metadata/stat_cache_test.go (hunk 9 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:539]: done

File: internal/cache/metadata/stat_cache_test.go (hunk 8 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache_test.go:443]: We could do that but it was making test less readable with multipl AAA blocks, addin that assert would be testing that first insert is working as expected which is already tested in another test where no entry is present, so did not add it in this test

File: internal/cache/metadata/stat_cache.go (hunk 6 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: renamed to sharedcachelookup.

Comment (vadlakondaswetha) [reply internal/cache/metadata/stat_cache.go]: I would prefer the other way. It would also make the caller job simpler.

func (sc *statCacheBucketView) LookUp(
	objectName string,
	now time.Time) (hit bool, m *gcs.MinObject) {
hit, entry := sc.sc.lruCacheLookup(folderName, now)
if hit {
m = entry.m
}
return
}

Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go]: done

File: internal/cache/metadata/stat_cache.go [hunk 11]
```diff
@@ -176,30 +201,86 @@ func (sc *statCacheBucketView) AddNegativeEntry(objectName string, expiration ti
...
-	now time.Time) (hit bool, m *gcs.MinObject) {
+	now time.Time) (bool, *gcs.MinObject) {
 	// Look up in the LRU cache.
-	value := sc.sharedCache.LookUp(sc.key(objectName))
+	hit, entry := sc.sharedcachelookup(objectName, now)
+	if hit {
+		return hit, entry.m
+	}
+
+	return false, nil
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:233]: Isn't the hit false here. It will be easy when you use named parameters for return types.
func (sc *statCacheBucketView) LookUp(
	objectName string,
	now time.Time) (hit bool, m *gcs.MinObject) {
hit, entry := sc.sharedCacheLookup(objectName, now)
if hit {
   m = entry.m
}

return

File: internal/cache/metadata/stat_cache.go [hunk 12]
```diff
@@ -176,30 +201,86 @@ func (sc *statCacheBucketView) AddNegativeEntry(objectName string, expiration ti
...
+	// Look up in the LRU cache.
+	hit, entry := sc.sharedcachelookup(folderName, now)
+	if hit {
+		return hit, entry.f
+	}
+
+	return false, nil
+}
+
+func (sc *statCacheBucketView) sharedcachelookup(key string, now time.Time) (bool, *entry) {
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:248]: camelCase for method names => sharedCacheLookup

File: internal/cache/metadata/stat_cache.go (hunk 11 above)
Comment (ankitaluthra1) [reply internal/cache/metadata/stat_cache.go:233]: line 230 hit is true and hit (true) is returned and line no 233 it should be false, seems working as expected, could please point which line no is returning incorrect value. Also should we avoid naked returns in all methods to be consistent, please see [ref](https://go.dev/tour/basics/7#:~:text=They%20can%20harm%20readability%20in%20longer%20functions.)

**Review by vadlakondaswetha — APPROVED** [review]
